### PR TITLE
Dynamic 2021 & Ghost Pool Protection

### PR DESCRIPTION
**Pull request by Xoxeyos**
# General Documentation

### Intent of your Pull Request

This ports over Dynamic 2021, and Ghost Pool Protection, which is sort of wrapped up with it, it splits the dynamic budget into two separate budgets, one for roundstart antagonists, and latejoin / midround antagonists, and a bunch of other changes. Here's the two PR's that are being ported. 
https://github.com/BeeStation/BeeStation-Hornet/pull/3797 - Dynamic 2021
https://github.com/BeeStation/BeeStation-Hornet/pull/3803 - Ghost Pool Protection

### Why is this change good for the game?

Dynamic uses less threat during the start of the round, and adds whatever is left over into the midround budget, certain events can hijack other random events replacing one event with another, say a sentience event with an alien infestation, of course relying on the budget. Highpop requirements have been removed, since they were sort of useless.

https://user-images.githubusercontent.com/35135081/109071468-9cab7e00-76a8-11eb-8f9f-2b920c602ef4.png
https://cdn.discordapp.com/attachments/712496193183154207/820869133494779904/unknown.png

# Wiki Documentation

### Briefly describe your PR and the impacts of it, in layman's terms. 
This will be the basis of the Wiki entry for your PR, and more information / detail is better for Wiki editors to integrate.

Dynamic will now split the threat into two separate budgets, one for roundstart, one for latejoiners and midrounders, more logging has been added for dynamic; The ghost pool protection allows admins to restrict what players can choose for ghostroles, say in case of an event, or to prevent griefing in some sort of way. The readme.md in game/gamemodes/dynamic/ is now alot more detailed.

### What should players be aware of when it comes to the changes your PR is implementing?

Dynamic is now split into two separate budgets, NT announcements regarding the threat level can be wrong, certain events can hijack other events.

### What general grouping does this PR fall under? 
Gamemode changes, dynamic update, admin logging and tools.

### Are there any aspects of the PR that you would like us not to mention on the Wiki?

Leaving the hijacking out and the threat level being faked might be nice, the ghost pool protection wouldn't important for anyone but admins.

### If there are any numerical values involved in your PR that will be relevant to a player, please note them here. 
I don't believe there are any.

# Changelog

:cl:  Xoxeyos, TheChosenEvilOne, Rohesie, IndieanaJones, Mothblocks, Qustinnus, tralezab, Theos
del: Removed high population override from dynamic.
fix: Dynamic can now be configured when made the forced secret mode.
tweak: dynamic threat reports are no longer perfectly accurate, they diverge and are sometimes downright wrong.
add: Dynamic now rolls a separate budget for round start and midround antagonists, meaning less round start antags will roll in favor of more midround ones.
admin: The antag_cap value now uses an equation that scales with population, rather than a fixed array.
admin: autotraitor_cooldown now uses deciseconds rather than ticks, so the older value of 450 now means 45 seconds. Set it to be 9000 if you want it to be accurate to the old value (15 minutes), or simply unset it.
admin: Midround/latejoin dynamic timers are now configurable.
admin: Syndicate Sleeper Agent will now give details when it fails.
fix: The autotraitor cooldown now respects configuration, rather than resetting back to 15 minutes after it is rolled.
fix: Fixed dynamic parameters not checking the configuration. This means that Manuel will now have slower midrounds, as was intended.
fix: Fixed multiple high impact rulesets being chosen on Dynamic.
fix: Midround dynamic injection no longer has a 100% chance of activating.
add: Dynamic will now play a part in controlling random event antagonists. If a dynamic midround injection is coming too soon, or too early, then the next midround injection will be buffed. Otherwise, it'll spawn normally.
admin: Admins can now cancel or replace midround rulesets.
fix: Admins can now correctly force midround rulesets.
admin: New Adminbus button- Ghost Pool Protection! It lets you enable and disable which sources ghosts can rejoin the round, minus their physical bodies getting revived.
/:cl:

### CODE REVIEW

**Comment by Xoxeyos** [comment]
A test merge would be needed for this, incase there wasn't something caught while testing it.

File: code/modules/mob/mob.dm [hunk 1]
```diff
@@ -63,12 +61,11 @@
   */
 /mob/Initialize()
 	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_MOB_CREATED, src)
-	GLOB.mob_list += src
-	GLOB.mob_directory[tag] = src
+	add_to_mob_list()
 	if(stat == DEAD)
-		GLOB.dead_mob_list += src
+		add_to_mob_list()
```
Comment (redmoogle) [inline code/modules/mob/mob.dm]:
```suggestion
		remove_from_alive_mob_list()
```

File: code/modules/mob/mob.dm [hunk 2]
```diff
@@ -63,12 +61,11 @@
...
 	else
-		GLOB.alive_mob_list += src
+		add_to_mob_list()
```
Comment (redmoogle) [inline code/modules/mob/mob.dm]:
```suggestion
		add_to_alive_mob_list()
```

**Review by redmoogle — DISMISSED** [review]
urgent fixes. bare minimum _append_ these changes

**Comment by Xoxeyos** [comment]
Thanks for fixing it, sorry I didn't get around to it.

**Comment by Xoxeyos** [comment]
I'm going to be removing the 'adminbus' tab and I'll be putting the ghost pool protection into the admin tab.

**Comment by Xoxeyos** [comment]
Traitors didn't load up when forced, I'll have to look into this further.

**Comment by Xoxeyos** [comment]
I'd like another test merge to be conducted again for this.

**Comment by Xoxeyos** [comment]
What the fuck is SOUND_ENVIROMENT

**Comment by redmoogle** [comment]
> 
> 
> What the fuck is SOUND_ENVIROMENT

https://github.com/yogstation13/Yogstation/pull/11106

**Comment by Xoxeyos** [comment]
current mob_defines does not have SOUND_ENVIROMENT_NONE in it.

**Comment by Xoxeyos** [comment]
This is ready for a test merge.

**Comment by Mothblocks** [comment]
*BeeStation's* Dynamic 2021?

**Comment by Xoxeyos** [comment]
further testing is required, all of the dynamic round last night, it only spawned infiltrators as latespawners, and one nonforced midround

**Comment by Xoxeyos** [comment]
Antags now spawn properly at the start of the round thanks to Jamie!

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 92e694acd1b4ddb5bdab7f42dbf75d255058fcd2
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: da56d524b85eeabf513b808c3ed6a8c9df605f9a

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 6eac6b8678728327d8878fa07e5c0f4655a8fa3e
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 92b18435786db50501d15b57a7d0f399158dcce7

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 6eac6b8678728327d8878fa07e5c0f4655a8fa3e
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 4cf045dcc80de3751ed501a3fdf71a161b4b6078

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 09dc7735c329443b1d7140a2d32a9b7e5388298f
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 8ce5531e4aa90f730f2881ca15125742069cc3b7

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 50e22d907e88fc4aed08712ae4939d0ca006b9bd
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: cfbef971e39f756f57c6351ccba16176fb7ca9b7

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: d244ed8af6b4625f69aff980ca6f669c8eab96cc
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: c9dd238bda1b13104771876231c7f061360f117a

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: b34fc836328b6ea784801f980d65536c09daad58
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 99a8136a2c3ea2386b7417cadce313eef7205037

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 51d15f34c48d85520d7724f0b4ab5cd650e8be05
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 08a5b01226a862a638855658a1fcb349f2180bdf

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: faf404bfec32c473db5ed3f6fa8061fcaa8b01d8
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 261ac2b415670d9615de140a821f0aed25f45412

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: dfdf9d25fba679e22d2ccaca56dc3ff10960e5af
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: a57b89a5cefc62a3059460567fc50d2b23efa5f5

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: dfdf9d25fba679e22d2ccaca56dc3ff10960e5af
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: d723368c137b9977e75b3e26dfb321c788297cdd

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 24d1f239ea0aa6c88caabde6243b7528fe97fb4c
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 1c84e2f8d3d922d24bcc92d7edb9ce9e32d5252e

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: e581b726a44943425eb38bb248e800e792d0fd9d
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: fc7122cf9f35688fd96d1b459cc1c9b61682e752

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 1befeb33968cd7fcd11341f081e90e149979098b
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 87f7ce98fbe6e2119e87704ba716dc7213e89bd9

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: bfcf8bfb1e922f1d6c8f6db19948a76bbfa39116
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: e56275b82d8a1b0a3a432a233629ac206864f83a

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 074dc717e2c59a75f46c232f5d068130b857d36d
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: e10404c9a10c2ee7f917b59885fa18ede7f0435a

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: aeb34a71e3fd3563bcfadbc47b1b7574c1f9359b
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 3fc96da3fe7b326106f45bec9fcd29d2afa4b3c4

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 5b96721222519db28316c53f605c7f8603965bfc
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 0e5c469765b215fb6c9c7efe7e82d2ab360f0212

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: a2b0d80bb2fe24c63b5611d324b01f653679b0fa
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 48ff5736e8e42a68fceb7e361203ebf0fa1bbeb2

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: 75e241882ae3eba86957aa8ef47647e805b2a28b
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: a573d4fda4232a30130441438b8607ed178df93e

**Comment by Yogbot-13** [comment]
#### Test Merge Updated

##### Server Instance
main

##### Merged By
JamieD12

##### Revision
Origin: b0848cac60e6718660d3154778f47593877aaf0e
Pull Request: 603b0473531283355ae824b81abc396632c02c8d
Server: 46c827dc9eeae1ddda6382961b987de97833563a

**Review by JamieD1 — APPROVED** [review]
